### PR TITLE
Refs #25406 -- Removed exception hiding in MySQL test database creati…

### PR DESCRIPTION
**Pull request by felixxm**
…on during `--keepdb`.

### CODE REVIEW

**Comment by adamchainz** [comment]
I might indeed :)

File: django/db/backends/mysql/creation.py [hunk 1]
```diff
@@ -17,24 +17,42 @@ def sql_table_creation_suffix(self):
...
+            'suffix': self.sql_table_creation_suffix(),
+        }
         with self._nodb_connection.cursor() as cursor:
             try:
-                cursor.execute("CREATE DATABASE %s" % qn(target_database_name))
+                self._execute_create_test_db(cursor, test_db_params, keepdb)
             except Exception:
-                if keepdb:
-                    return
                 try:
```
Comment (charettes) [inline django/db/backends/mysql/creation.py]: How can we end up in this branch if `_execute_create_test_db` catches all exceptions?

File: django/db/backends/mysql/creation.py [hunk 2]
```diff
@@ -17,24 +17,42 @@ def sql_table_creation_suffix(self):
...
+    def _execute_create_test_db(self, cursor, parameters, keepdb=False):
+        try:
+            if keepdb:
+                # If the database should be kept, add "IF NOT EXISTS" to avoid
+                # "database exists" error, also temporarily disable "database
+                # exists" warning.
+                cursor.execute('''
+                    SET sql_notes = 0;
+                    CREATE DATABASE IF NOT EXISTS %(dbname)s %(suffix)s;
+                    SET sql_notes = 1;
```
Comment (adamchainz) [inline django/db/backends/mysql/creation.py]: what if it was set to 0 by connection init or server param? We should just reset it back to the original value...

```sql
SET @_tmp_sql_notes := @@sql_notes;
SET sql_notes = 0;
CREATE DATABASE IF NOT EXISTS ...
SET sql_notes = @_tmp_sql_notes;
```

Comment (adamchainz) [inline django/db/backends/mysql/creation.py]: Actually that can be one line `SET @_tmp_sql_notes := @@sql_notes, sql_notes = 0;`

File: django/db/backends/mysql/creation.py (hunk 1 above)
Comment (felixxm) [inline django/db/backends/mysql/creation.py]: You're right. I fixed this.

**Comment by felixxm** [comment]
@adamchainz, @charettes, @timgraham Thanks for review. I updated PR.

File: django/db/backends/mysql/creation.py (hunk 1 above)
Comment (charettes) [inline django/db/backends/mysql/creation.py]: Did you add a test case for it as well? I don't see anything passing `1007` in there?

Comment (felixxm) [inline django/db/backends/mysql/creation.py]: Done. I forgot to commit tests.

File: tests/backends/test_creation.py [hunk 3]
```diff
@@ -191,3 +193,34 @@ def test_create_test_user(self, *mocked_objects):
...
+        # Simulate test database creation raising "database exists"
+        with self.patch_test_db_creation(self._execute_raise_database_exists):
+            with mock.patch('builtins.input', return_value='no'):
+                with self.assertRaises(SystemExit):
+                    # SystemExit is raised if the user answers "no" to the
+                    # prompt asking if it's okay to delete the test database.
+                    creation._create_test_db(verbosity=0, autoclobber=False, keepdb=False)
+            # "Database exists" shouldn't appear when keepdb is on
+            creation._create_test_db(verbosity=0, autoclobber=False, keepdb=True)
+        # Simulate test database creation raising unexpected error
```
Comment (adamchainz) [inline tests/backends/test_creation.py]: these are two tests and should be two test methods e.g. `test_db_database_exists` and `test_access_denied`, then you don't need comments breaking them up

**Pull request merged**